zkevm-circuits/src/evm_circuit/execution/balance.rs

```diff
@@ -0,0 +1,313 @@
+use crate::evm_circuit::execution::ExecutionGadget;
+use crate::evm_circuit::param::N_BYTES_ACCOUNT_ADDRESS;
+use crate::evm_circuit::step::ExecutionState;
+use crate::evm_circuit::util::common_gadget::SameContextGadget;
+use crate::evm_circuit::util::constraint_builder::Transition::Delta;
+use crate::evm_circuit::util::constraint_builder::{
+    ConstraintBuilder, ReversionInfo, StepStateTransition,
+};
+use crate::evm_circuit::util::{from_bytes, select, CachedRegion, Cell, RandomLinearCombination};
+use crate::evm_circuit::witness::{Block, Call, ExecStep, Rw, Transaction};
+use crate::table::{AccountFieldTag, CallContextFieldTag};
+use crate::util::Expr;
+use eth_types::evm_types::GasCost;
+use eth_types::{Field, ToAddress, ToLittleEndian};
+use halo2_proofs::circuit::Value;
+use halo2_proofs::plonk::Error;
+
+#[derive(Clone, Debug)]
+pub(crate) struct BalanceGadget<F> {
+    same_context: SameContextGadget<F>,
+    address: RandomLinearCombination<F, N_BYTES_ACCOUNT_ADDRESS>,
+    reversion_info: ReversionInfo<F>,
+    tx_id: Cell<F>,
+    is_warm: Cell<F>,
+    balance: Cell<F>,
+    exists: Cell<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for BalanceGadget<F> {
+    const NAME: &'static str = "BALANCE";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::BALANCE;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let address = cb.query_rlc();
+        cb.stack_pop(address.expr());
+
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+        let mut reversion_info = cb.reversion_info_read(None);
+        let is_warm = cb.query_bool();
+        cb.account_access_list_write(
+            tx_id.expr(),
+            from_bytes::expr(&address.cells),
+            1.expr(),
+            is_warm.expr(),
+            Some(&mut reversion_info),
+        );
+
+        let exists = cb.query_bool();
+        let balance = cb.condition(exists.expr(), |cb| {
+            let balance = cb.query_cell();
```

Comment:
can the "query_cell" be moved out of the closure? It is some "structure/allocation" info, better not be guarded with "if" branch.

Reply:
Moved out `query_cell` as:
```
let balance = cb.query_cell();
let exists = cb.query_bool();
cb.condition(exists.expr(), |cb| {
    cb.account_read(
        from_bytes::expr(&address.cells),
        AccountFieldTag::Balance,
        balance.expr(),
    );
});
```